docs/decisions/20240604-hdkey-domain-separation

```diff
@@ -0,0 +1,62 @@
+# HDKey determinstic key and domain derivation when using different algorithms
+
+- Status: in-progress 
+- Deciders: Javi, Yurii, Jesus, Matt, Ezequiel
+- Date: 2024-06-04
+- Tags: Domain separation, Deterministic wallets
+
+
+## Context and Problem Statement
+
+The main question to answer is: How can we add ed25519 HDKey deterministic wallet implementation on top of existing secp256k1 HDKey implementation and make it imposssible for the users to generate the same keys using the same seed?
+
+For broader context, we are now introducing ed25519 HDKey deterministic wallet with Cardanos bip32 implementation for ed25519 and this will work together with our existing standard bip32 implementation which is used by secp256k1.
+
+This deterministic keys are going to be used by the issuer to issue credentials with both algorithms and it is important to mention, right now the Bip32 is deriving using the following paths.
+
+Currently in Bip32 (secp256k1), the path looks like this m/0'/0'/0' and we are just incrementing the accountIndex whenever it is needed.
+There is a risk now where user's could potentially be creating the same keys using different key curves and find a colision. 
+
+
+## Decision Drivers
+
+A desired solution should
+
+- Ensure there's enough domain separation between the 2 types of keys so that they could never be generating the same keys.
+- This solution or any solition we build to change or improve derivationPaths will require to have backwards compatibility so any user could still import the keys with the old paths.
+- The proposed solution must also be future proof and secure enough to use.
+- Go through a proper security audit and require some i+d to we are not going to have any risk.
+
+## Considered Options
+
+We have considered several approaches, initially changing the derive hashing function to just include the text "ed25519", we also explored changing the derivationPath to ensure the rotation and domain separation
+
+## Decision Outcome
+
+After a careful analysis we concluded the following points:
+- Changing the hashing function would prevent other implementations from being able to restore the same derived keys, as the hashing would not be standard.
+- Defining a proper derivation path would allow us to rotate and add domain separation
+- DerivationPath is not something we cannot recover or move from, in case we wanted to improve derivation paths, we MUST keep backwards compatibility and always allow user's to restore those keys with the correct derivationPath if they required to do so.
```

Comment:
my understanding is that the seed alone is likely not enough to do a recovery of users state. There is metadata (e.g. what DIDs are unpublished, which are published, which is the current status of the DID, etc) that is (should be) part of a recovery process.
In such a case, it is not too much work to add derivation paths to that metadata if needed. So, I wouldn't be too concern of changing the path

Reply:
Yes correct it would not be enough but as we are just increasing indexes there could be a recovery process as u say where you could run like discovery, etc.

Yup, agreed